DAOS-7963 Java Client build failing on Fedora

zjf2012 opened the pull request:
remove the unnecessary import of "com.sun.corba.se.impl.orbutil.concurrent" which failed Java code compilation in Fedora.

Signed-off-by: jiafu zhang <jiafu.zhang@intel.com>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

ashleypittman commented:
If this is fixed then you can uncomment the lines linked below which will enable testing of this build for every PR.  It makes more sense to enable the test in the same PR as the fix rather than to test manually and have a second PR to enable the test. 

https://github.com/daos-stack/daos/blob/master/.github/workflows/ci.yml#L76-L80

zjf2012 commented:
> If this is fixed then you can uncomment the lines linked below which will enable testing of this build for every PR. It makes more sense to enable the test in the same PR as the fix rather than to test manually and have a second PR to enable the test.
> 
> https://github.com/daos-stack/daos/blob/master/.github/workflows/ci.yml#L76-L80

ok, it should be fixed. Let me try.

ashleypittman commented:
This has failed the build, there appear to be (possibly longstanding) compiler warnings at https://github.com/daos-stack/daos/pull/6716/checks?check_run_id=3607789781#step:6:9383 which aren't fatal but look like they should be, and test failures at https://github.com/daos-stack/daos/pull/6716/checks?check_run_id=3607789781#step:6:10486 which are.

ashleypittman commented on .github/workflows/ci.yml:
```diff
@@ -73,11 +73,11 @@ jobs:
+      run: docker build . -f utils/docker/Dockerfile.centos.8
+                          --build-arg COMPILER=${{ matrix.compiler }}
+                          --build-arg DEPS_JOBS=8
+                          --build-arg BASE_DISTRO=fedora:34
```
FYI - you can test this code locally by running the command from here:

docker build . -f utils/docker/Dockerfile.centos.8 --build-arg BASE_DISTRO=fedora:34

If you don't have a docker server setup I can let you access one of mine on the wolf cluster.

zjf2012 replied:
that would be great. please share me your docker server in wolf. thanks.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

zjf2012 commented:
@ashleypittman finally, the issue is fixed. But I added a new dockerfile for Fodera. Please let me know if it's inappropriate.  The goal is to not install maven with JDK-11 dependencies from dnf repo. Instead, I download and install only maven by myself. And it worked with JDK-8.

ashleypittman commented:
> @ashleypittman finally, the issue is fixed. But I added a new dockerfile for Fodera. Please let me know if it's inappropriate. The goal is to not install maven with JDK-11 dependencies from dnf repo. Instead, I download and install only maven by myself. And it worked with JDK-8.

Yes it is appropriate, there are other PRs doing the same thing.  What you should do however is change the dockerfile name in all parts of the ci.yaml file, and remove the BASE_DISTRO code, and change "FROM" in the fedora dockerfile.

We should try and find a way of not hardcoding version numbers in the Fedora build, is there not a repo that we can install from at all?

Happy to work with you on these two items to get it resolved.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

zjf2012 commented:
> > @ashleypittman finally, the issue is fixed. But I added a new dockerfile for Fodera. Please let me know if it's inappropriate. The goal is to not install maven with JDK-11 dependencies from dnf repo. Instead, I download and install only maven by myself. And it worked with JDK-8.
> 
> Yes it is appropriate, there are other PRs doing the same thing. What you should do however is change the dockerfile name in all parts of the ci.yaml file, and remove the BASE_DISTRO code, and change "FROM" in the fedora dockerfile.
> 
> We should try and find a way of not hardcoding version numbers in the Fedora build, is there not a repo that we can install from at all?
> 
> Happy to work with you on these two items to get it resolved.

Just updated code accordingly. For the maven repo hard coding, I tried to get maven from Fedora repo. But maven gets installed with JDK-11 as dependency. I tried to suppress JDK-11 and configure Java to use JDK-8 with update-alternative. However, it's failed. So, I download only maven from maven site. 

I've add an arg, MAVEN_VERSION, so that we can configure maven version later.

ashleypittman commented on .github/workflows/ci.yml:
```diff
@@ -62,33 +62,28 @@ jobs:
         submodules: true
     - name: Prepare base image in Docker
-      run: docker build . -f utils/docker/Dockerfile.centos.8
+      run: docker build . -f utils/docker/Dockerfile.fodera
```
Spelling mistake.  You'll need to modify this in lots of places and rename the file.
```suggestion
      run: docker build . -f utils/docker/Dockerfile.fedora
```
Although now fedora is in it's own file you can collapse this a bit more, add ", fedora" to the matix azis on L18 and remove Lines 51 and onwards in this file.

ashleypittman commented on utils/docker/Dockerfile.centos.8:
```diff
@@ -307,6 +307,10 @@ RUN echo -e "<settings>\n\
 
 ARG DAOS_JAVA_BUILD=$DAOS_BUILD
 
+RUN CHOICE=$(echo "" | sudo alternatives --config java | grep "java-1.8" | tr -cd '[[:digit:]]') && \
+        echo ${CHOICE:0:1} | sudo alternatives --config java
+RUN java -version
+
```
Should this still be here for CentOS 8?

You should remove the BASE_DISTRO arg from the top of this file however as it's no longer used.

ashleypittman commented on utils/docker/Dockerfile.fodera:
```diff
@@ -0,0 +1,324 @@
+#
+
+# Pull base image
+ARG BASE_DISTRO=fedora:34
+FROM $BASE_DISTRO
```
Unless this is going to be modified on the command line you don't need to do this.
```suggestion
FROM fedora:34
```

ashleypittman commented on utils/docker/Dockerfile.fodera:
```diff
@@ -0,0 +1,324 @@
+ARG MAVEN_VERSION=3.6.3
+USER root:root
+
+RUN [ "$DAOS_JAVA_BUILD" != "yes" ] || ( wget https://archive.apache.org/dist/maven/maven-3/$MAVEN_VERSION/binaries/apache-maven-$MAVEN_VERSION-bin.tar.gz && tar -xf apache-maven-$MAVEN_VERSION-bin.tar.gz && find ./ -name "mvn" | { read d; f=$(dirname $d); cp -r $(readlink -f $f)/../* /usr/local; } )
```
Is there no way at all to install the correct package, even if it would mean modifying the yum config to do it?

johannlombardi commented:
What is the status of this patch? Shall we land it? Please advise.

ashleypittman commented:
> What is the status of this patch? Shall we land it? Please advise.

WIP, but we should land it once complete.

ashleypittman commented:
This will need updating unfortunately, we have an ongoing discussion about if we spilt out the dockerfiles like this or if we keep them as one and control them with args, as you can see there's a lot of duplication if we split them and that is going to significantly increase the ongoing support burden.

Ping me on IM and we can have a look at the best way forward.

johannlombardi commented:
This has been fixed already? Can we close?

zjf2012 commented:
> This has been fixed already? Can we close?

yes, it's fixed. But @ashleypittman seems having some concerns about the new fedora ci file. He said he'll talk to me later.

ashleypittman commented:
No, it's not fixed.  Java still does not build on Fedora that I'm aware of.

zjf2012 commented:
> No, it's not fixed. Java still does not build on Fedora that I'm aware of.

so, I should merge my fix to your latest ci adjustment?

ashleypittman commented:
> > No, it's not fixed. Java still does not build on Fedora that I'm aware of.
> 
> so, I should merge my fix to your latest ci adjustment?

Yes, but your fix is to split the centos 8 docker file into multiple files per distro which is something that we were considering but have decided to avoid if we can, so ideally if we still need different packages for different distros which are built from the same dockerfile then we'll probably need to put a conditional around the install command.

zjf2012 commented:
> > > No, it's not fixed. Java still does not build on Fedora that I'm aware of.
> > 
> > 
> > so, I should merge my fix to your latest ci adjustment?
> 
> Yes, but your fix is to split the centos 8 docker file into multiple files per distro which is something that we were considering but have decided to avoid if we can, so ideally if we still need different packages for different distros which are built from the same dockerfile then we'll probably need to put a conditional around the install command.

ok, I'll try to make it conditional.

daosbuild1 commented on utils/scripts/install-centos8.sh:
```diff
@@ -68,4 +69,12 @@ dnf install \
+
+if [ "$fedora_java" = "yes" ]; then
+	dnf install java-1.8.0-openjdk-devel
+	curl --output ./apache-maven-3.6.3-bin.tar.gz https://archive.apache.org/dist/maven/maven-3/3.6.3/binaries/apache-maven-3.6.3-bin.tar.gz && tar -xf apache-maven-3.6.3-bin.tar.gz && find ./ -name "mvn" | { read d; f=$(dirname $d); cp -r $(readlink -f $f)/../* /usr/local; }
```
(style)  line over 100 characters
(lint) read without -r will mangle backslashes. [SC2162]
(lint) Double quote to prevent globbing and word splitting. [SC2086]
(lint) Quote this to prevent word splitting. [SC2046]
(lint) Double quote to prevent globbing and word splitting. [SC2086]

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6716/16/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6716/16/execution/node/88/log

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6716/22/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-6716/22/execution/node/113/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

brianjmurrell commented on .github/workflows/ci2.yml:
```diff
@@ -72,7 +72,7 @@ jobs:
             base: centos.8
             with: fedora:35
-            java: no
+            java: yes
```
Since this ends up getting evaluated by shell script, use `true`(/`false`) here as it lets you write more language natural shell script later.

zjf2012 replied:
sounds good.

brianjmurrell commented on utils/docker/Dockerfile.centos.8:
```diff
@@ -105,11 +105,20 @@ gpgcheck=False\n" >> /etc/yum.repos.d/$repo:$branch:$build_number.repo;   \
 RUN chmod +x /tmp/install.sh
-RUN /tmp/install.sh
+
+RUN if [[ "$BASE_DISTRO" = fedora* ]] && [[ "$DAOS_JAVA_BUILD" == "yes" ]]; then \
```
With the previous suggestion this can become:

```suggestion
RUN if [[ "$BASE_DISTRO" = fedora* ]] && "$DAOS_JAVA_BUILD"; then \
```

brianjmurrell commented on utils/docker/Dockerfile.centos.8:
```diff
@@ -105,11 +105,20 @@ gpgcheck=False\n" >> /etc/yum.repos.d/$repo:$branch:$build_number.repo;   \
-RUN /tmp/install.sh
+
+RUN if [[ "$BASE_DISTRO" = fedora* ]] && [[ "$DAOS_JAVA_BUILD" == "yes" ]]; then \
+	. /tmp/install.sh yes; 	\
+    else 			\
+	. /tmp/install.sh no; 	\
+    fi
```
Again, use `true`/`false` for shell script booleans.

```suggestion
	. /tmp/install.sh true; 	\
    else 			\
	. /tmp/install.sh false; 	\
    fi
```

brianjmurrell commented on utils/docker/Dockerfile.centos.8:
```diff
@@ -215,7 +224,7 @@ ARG COMPILER=gcc
 ARG DAOS_BUILD_TYPE=$DAOS_TARGET_TYPE
 ARG DAOS_BUILD=$DAOS_DEPS_BUILD
-
+RUN echo "DAOS_BUILD is $DAOS_BUILD"
```
Is the debug that should be removed?

zjf2012 replied:
removed

brianjmurrell commented on utils/scripts/install-centos8.sh:
```diff
@@ -67,5 +68,20 @@ dnf install \
+    which \
     yasm
+
+echo "is fedora java? $fedora_java"
```
More debug that needs removing?

zjf2012 replied:
removed

brianjmurrell commented on utils/scripts/install-centos8.sh:
```diff
@@ -67,5 +68,20 @@ dnf install \
+
+echo "is fedora java? $fedora_java"
+
+if [ "$fedora_java" = "yes" ]; then
```
Per above, use natural shell booleans:

```suggestion
if "$java"; then
```

brianjmurrell commented on utils/scripts/install-centos8.sh:
```diff
@@ -10,6 +10,8 @@
 
 set -e
 
+fedora_java=$1
```
This isn't really anything to do with _Fedora_ is it?  It's just whether to install dependencies for the Java client build, right?  Even though currently we only do that on Fedora, it could be done on other platforms also, right?

```suggestion
java=$1
```

zjf2012 replied:
yes. it's really about fedora. Java-11 is installed along with maven automatically if we install maven from repo in fedora. I tried several ways to remove the Java-11 dep but failed. Thus, I need to handle fedora separately. 

Since you mentioned no Java and maven at all if it builds without Java. I split "fedora_java" to two variables, "java" and "fedora".

brianjmurrell commented on utils/scripts/install-centos8.sh:
```diff
@@ -67,5 +68,20 @@ dnf install \
+
+if [ "$fedora_java" = "yes" ]; then
+	dnf install java-1.8.0-openjdk-devel
+	curl --output ./apache-maven-3.6.3-bin.tar.gz \
+https://archive.apache.org/dist/maven/maven-3/3.6.3/binaries/apache-maven-3.6.3-bin.tar.gz \
+&& tar -xf apache-maven-3.6.3-bin.tar.gz \
+&& find ./ -name "mvn" | \
+{ read -r d; f="$(dirname "$d")"; cp -r "$(readlink -f "$f")"/../* /usr/local; }
```
Why are we doing this for Fedora but using the `maven` RPM for other distros?

zjf2012 replied:
see above comments

brianjmurrell commented on utils/scripts/install-centos8.sh:
```diff
@@ -67,5 +68,20 @@ dnf install \
+&& tar -xf apache-maven-3.6.3-bin.tar.gz \
+&& find ./ -name "mvn" | \
+{ read -r d; f="$(dirname "$d")"; cp -r "$(readlink -f "$f")"/../* /usr/local; }
+else
+	dnf install maven
```
Is `maven` even needed if the Java client is not being built?

zjf2012 replied:
adjusted the variable. see above comments.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Small completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6716/29/display/redirect

ashleypittman commented on .github/workflows/ci2.yml:
```diff
@@ -68,15 +68,15 @@ jobs:
             base: centos.8
             with: rockylinux/rockylinux:8
-            java: yes
+            java: true
```
With the PR we should be building Java for all builds so it no longer needs to be an matrix index, you can remove these lines.

ashleypittman commented on .github/workflows/ci2.yml:
```diff
@@ -110,8 +110,8 @@ jobs:
                           --build-arg DEPS_JOBS
-                          --build-arg BASE_DISTRO
-                          --build-arg DAOS_JAVA_BUILD
+                          --build-arg BASE_DISTRO=${{ matrix.with }}
```
If no value is given here then the value from the environment will be used, which is set on L82 so the change on this line isn't required.

ashleypittman commented on .github/workflows/ci2.yml:
```diff
@@ -110,8 +110,8 @@ jobs:
-                          --build-arg BASE_DISTRO
-                          --build-arg DAOS_JAVA_BUILD
+                          --build-arg BASE_DISTRO=${{ matrix.with }}
+                          --build-arg DAOS_JAVA_BUILD=${{ matrix.java }}
```
You can set this to yes here directly.  It should be yes rather than true.

ashleypittman commented on utils/docker/Dockerfile.centos.8:
```diff
@@ -105,11 +105,20 @@ gpgcheck=False\n" >> /etc/yum.repos.d/$repo:$branch:$build_number.repo;   \
         dnf repolist; \
         dnf --disablerepo=\* --enablerepo=$repo:$branch:$build_number makecache; \
     done
+# For conditional build DAOS and Java
+ARG DAOS_BUILD=$DAOS_DEPS_BUILD
+ARG DAOS_JAVA_BUILD=$DAOS_BUILD
+ARG BASE_DISTRO
```
Putting it this early in the dockerfile will mean that changing if java is built or not will re-install all the packages and rebuild all the dependencies, it would be better to have the customisations (or at least the DAOS_JAVA_BUILD ones) later on, preferably after L245.

ashleypittman reviewed (commented):
This will need changes to landing-builds as well as ci2 to account for master builds.

If you copy the trigger changes from https://github.com/daos-stack/daos/pull/7677/files to here then landing builds will also trigger on this PR so you can verify them there.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage NLT completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-6716/33/display/redirect

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

ashleypittman commented on utils/docker/Dockerfile.centos.8:
```diff
@@ -243,7 +243,18 @@ RUN echo -e "<settings>\n\
 </settings>" > /home/daos/.m2/settings.xml
 
 ARG DAOS_JAVA_BUILD=$DAOS_BUILD
+ARG BASE_DISTRO
```
You already have ARG BASE_DISTRO above, right at the start so don't need this line.

zjf2012 replied:
As tested, I cannot read any value from BASE_DISTRO if I don't redefine it again. I don't know why yet. Any idea?

ashleypittman replied:
No, as far as I know it's just a regular arg.

zjf2012 replied:
As tested again, I need to redefine this arg to make fedora build work.

ashleypittman commented on .github/workflows/ci2.yml:
```diff
@@ -68,21 +68,18 @@ jobs:
       DOCKER_BASE: ${{ matrix.base }}
       COMPILER: ${{ matrix.compiler }}
-      DAOS_JAVA_BUILD: ${{ matrix.java }}
+      DAOS_JAVA_BUILD: yes
```
As this is set on L111 you don't need it in the env now.

zjf2012 replied:
> did you see my earlier comment about needing to update the landings triggers as well?

I'll do it probably tomorrow.

ashleypittman commented on utils/scripts/install-centos8-java.sh:
```diff
@@ -0,0 +1,24 @@
+#!/bin/sh
```
```suggestion
#!/bin/sh -e
```

ashleypittman commented on utils/scripts/install-centos8.sh:
```diff
@@ -67,5 +65,6 @@ dnf install \
     python3-yaml \
     sg3_utils \
     valgrind-devel \
+    which \
```
Is this change needed?

ashleypittman replied:
Just checking it was intentional.  Best to leave it here and have the minimal set in the conditional script.

ashleypittman reviewed (commented):
did you see my earlier comment about needing to update the landings triggers as well?

ashleypittman commented:
Next time you push this PR can you add -ntp to the mvn commands in the dockerfiles?

```
@@ -207,7 +207,7 @@ RUN echo -e "<settings>\n\
 ARG DAOS_JAVA_BUILD=$DAOS_BUILD

 RUN [ "$DAOS_JAVA_BUILD" != "yes" ] || \
-    mvn clean install -T 1C -DskipITs -Dgpg.skip -Ddaos.install.path=/opt/daos
+    mvn clean install -ntp -T 1C -DskipITs -Dgpg.skip -Ddaos.install.path=/opt/daos
 WORKDIR /home/daos

 ARG DAOS_KEEP_SRC=no
````

zjf2012 commented:
> Next time you push this PR can you add -ntp to the mvn commands in the dockerfiles?
> 
> ```
> @@ -207,7 +207,7 @@ RUN echo -e "<settings>\n\
>  ARG DAOS_JAVA_BUILD=$DAOS_BUILD
> 
>  RUN [ "$DAOS_JAVA_BUILD" != "yes" ] || \
> -    mvn clean install -T 1C -DskipITs -Dgpg.skip -Ddaos.install.path=/opt/daos
> +    mvn clean install -ntp -T 1C -DskipITs -Dgpg.skip -Ddaos.install.path=/opt/daos
>  WORKDIR /home/daos
> 
>  ARG DAOS_KEEP_SRC=no
> ```

sure

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/node_local_test.py:192](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L192):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:400](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L400):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:455](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L455):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:459](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L459):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:466](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L466):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:584](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L584):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:586](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L586):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:610](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L610):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:641](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L641):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:653](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L653):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:668](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L668):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:750](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L750):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1000](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1000):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1001](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1001):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1066](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1066):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1109](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1109):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1703](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1703):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1708](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1708):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1714](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1714):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1743](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1743):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1775](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1775):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1776](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1776):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1789](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1789):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1804](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1804):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1883](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1883):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1943](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1943):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1968](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1968):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1985](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1985):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1986](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L1986):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2038](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L2038):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2199](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L2199):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2326](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L2326):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2442](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L2442):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2556](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L2556):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2590](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L2590):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2677](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L2677):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2678](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L2678):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2701](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L2701):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2702](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L2702):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2944](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L2944):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:3141](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L3141):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:3170](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L3170):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:3193](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L3193):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:3226](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L3226):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:3395](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L3395):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:3529](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L3529):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:3634](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L3634):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:4001](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L4001):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:4002](https://github.com/daos-stack/daos/blob/87c85aa5be9294ea42701a413d905741cf6e94bc/utils/node_local_test.py#L4002):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

zjf2012 commented:
> > Next time you push this PR can you add -ntp to the mvn commands in the dockerfiles?
> > ```
> > @@ -207,7 +207,7 @@ RUN echo -e "<settings>\n\
> >  ARG DAOS_JAVA_BUILD=$DAOS_BUILD
> > 
> >  RUN [ "$DAOS_JAVA_BUILD" != "yes" ] || \
> > -    mvn clean install -T 1C -DskipITs -Dgpg.skip -Ddaos.install.path=/opt/daos
> > +    mvn clean install -ntp -T 1C -DskipITs -Dgpg.skip -Ddaos.install.path=/opt/daos
> >  WORKDIR /home/daos
> > 
> >  ARG DAOS_KEEP_SRC=no
> > ```
> 
> sure

"-ntp" is only for maven 3.6.1 and newer version. I need to use "-B -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn" in centos7 and centos8.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

FYI: Errors found in lines not modified in the patch:

[utils/node_local_test.py:192](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L192):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:400](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L400):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:455](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L455):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:459](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L459):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:466](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L466):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:584](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L584):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:586](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L586):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:610](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L610):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:641](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L641):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:653](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L653):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:668](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L668):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:750](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L750):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1000](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1000):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1001](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1001):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1066](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1066):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1109](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1109):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1703](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1703):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1708](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1708):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1714](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1714):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1743](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1743):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1775](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1775):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1776](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1776):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1789](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1789):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1804](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1804):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1883](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1883):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1943](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1943):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1968](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1968):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1985](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1985):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:1986](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L1986):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2038](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L2038):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2199](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L2199):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2326](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L2326):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2442](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L2442):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2556](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L2556):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2590](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L2590):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2677](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L2677):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2678](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L2678):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2701](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L2701):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2702](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L2702):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:2944](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L2944):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:3141](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L3141):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:3170](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L3170):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:3193](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L3193):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:3226](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L3226):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:3395](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L3395):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:3529](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L3529):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:3634](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L3634):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:4001](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L4001):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

[utils/node_local_test.py:4002](https://github.com/daos-stack/daos/blob/560d2f4281f5d5de2d7887db8569034061b8c5b5/utils/node_local_test.py#L4002):
(pylint-consider-using-with) Consider using 'with' for resource-allocating operations

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

ashleypittman commented on utils/scripts/install-centos8-java.sh:
```diff
@@ -0,0 +1,22 @@
+
+fedora=$1
+
+if "$fedora"; then
```
As discussed on teams I think "dnf install maven-openjdk8" would work here, and potentially you could do the following, and have it work on any distro without reading $1 or parsing the BASE_DISTRO variable.

```dnf install maven-openjdk8 || dnf install maven```

or

```
if [ -e /etc/fedora-release]
then
    dnf install maven-openjdk8
else
   dnf install maven
fi

dnf clean all

zjf2012 replied:
sounds good. let me try.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

ashleypittman commented on utils/scripts/install-centos8-java.sh:
```diff
@@ -0,0 +1,14 @@
+# installed specifically.
+# For centos, we can install both java-8 and maven from repo.
+
+if [ -e /etc/fedora-release ]; then
+	dnf install java-1.8.0-openjdk java-1.8.0-openjdk-devel maven-openjdk8
+else
+        dnf install java-1.8.0-openjdk maven
+fi
```
I think at this point you can just put it in the initial install-centos8.sh, and not have the second script.
```suggestion
if [ -e /etc/fedora-release ]; then
	dnf install java-1.8.0-openjdk java-1.8.0-openjdk-devel maven-openjdk8
else
        dnf install java-1.8.0-openjdk maven
fi
```

zjf2012 replied:
It's not only for fedora, but also conditional Java build. check line 247 in dockerfile.centos.8.  

"RUN if [[ "$DAOS_JAVA_BUILD" == "yes" ]]; then "

I need to introduce a parameter to install-centos8.sh to check if it's java build.

ashleypittman replied:
In all other cases we take the approach that we install all needed components/rpms at the beginning, regardless of if they'll be used or not.  As discussed adding additional parameters will break the docker caching of images and cause the dependencies to be rebuilt unnecessarily, so IMO we should install these packages unconditionally earlier in the process.

What you say does make sense that we should not install them until we need them, and then only if requested however we also build daos and the java client for every PR, and that means loading cached docker images from master that avoids downloading updates however with this code as it is then each PR would need to download the maven packages twice (once for gcc, once for clang) for each distro, and this introduces additional network load, but crucially an additional point of failure.

zjf2012 replied:
make sense. Let me change it.

ashleypittman commented on utils/scripts/install-centos8.sh:
```diff
@@ -68,3 +67,15 @@ dnf --nodocs install \
+        dnf install maven
+fi
+
+dnf clean all
```
This line isn't needed now, the dockerfile will do this after calling the scipt.

zjf2012 replied:
removed

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.
